packages/malloc/src/pool.ts

```diff
@@ -307,3 +368,60 @@ export class MemPool implements IMemPool {
         block.next = ptr;
     }
 }
+
+
+const MEM_BLOCK_SIZE = 0;
+const MEM_BLOCK_NEXT = 1;
+
+const SIZEOF_MEM_BLOCK = 2 * Uint32Array.BYTES_PER_ELEMENT;
+
+/**
+ * All of the addresses that saved are the memory address, and not the data address.
+ * The user see the data address, we need to convert them in and out
+ */
+class MemBlockWrapper implements MemBlock {
+    public static OVERHEAD = SIZEOF_MEM_BLOCK;
+
+    public static toDataAddress(blockAddress: number) {
+        return blockAddress + SIZEOF_MEM_BLOCK;
+    }
+
+    public static toMemoryAddress(dataAddress: number) {
+        return dataAddress - SIZEOF_MEM_BLOCK;
+    }
+
+    constructor(
+        private u32: Uint32Array, 
+        /**
+         * This is the memory address, not the data address
+         */
+        public readonly addr: number) {}
+
+    public get next(): MemBlock | null {
+        if (this.nextAddress !== 0) {
+            return new MemBlockWrapper(this.u32, this.nextAddress);
+        }
+
+        return null;
+    }
+
+    public set next(value: MemBlock | null) {
+        this.nextAddress = value ? value.addr : 0;
+    }
+
+    public get size() {
+        return this.u32[((this.addr - this.u32.byteOffset) >> 2) + MEM_BLOCK_SIZE];
```

Comment:
@postspectacular
tradeoff. to have offset, or not ?

Reply:
Why is this offset needed here? I assumed that `u32` is a view of the entire array buffer (just like `u8`) and so the offset is always 0, no?